### PR TITLE
Gemini Nod+Shuffle

### PR DESCRIPTION
**Pull request by profxj**
Introduces an approach (non-optimal) for Gemini GMOS Nod+Shuffle observations.

Also improves all of the GMOS wavelength solutions.

### CODE REVIEW

File: CHANGES.rst [hunk 1]
```diff
@@ -45,6 +45,8 @@
   alignment boxes from slitmask design.
 - Added a table with DEIMOS slitmask design and objects info to the
   SlitTraceSet datamodel
+- Incorporates a procedure to enable GMOS N-S observations
```
Comment (jhennawi) [inline CHANGES.rst:48]: N-S is ambiguous given North/South. Can you write out nod and shuffle.

File: pypeit/core/wavecal/wvutils.py [hunk 2]
```diff
@@ -5,10 +5,13 @@
 .. include:: ../include/links.rst
 
 """
+import os
 import numpy as np
 import os
 import numba as nb
```
Comment (jhennawi) [inline pypeit/core/wavecal/wvutils.py:11]: Is the number import being used? I sort of doubt it. I would comment this out.

File: pypeit/wavecalib.py [hunk 3]
```diff
@@ -117,6 +117,18 @@ def _bundle(self):
     @classmethod
     def _parse(cls, hdu, ext=None, transpose_table_arrays=False, debug=False,
                hdu_prefix=None):
+        """
+
+        Args:
```
Comment (jhennawi) [inline pypeit/wavecalib.py:122]: Please fill in the docs.

File: doc/gemini_gmos.rst [hunk 4]
```diff
@@ -10,6 +10,14 @@ This file summarizes several instrument specific
 settings that are related to the Gemini/GMOS spectrograph.
 
 
+Nod and Shuffle
```
Comment (jhennawi) [inline doc/gemini_gmos.rst:13]: Can we add a GitHub issue saying that we need to eventually do this right.

**Review by jhennawi — CHANGES_REQUESTED** [review]
Minor changes requested.

File: CHANGES.rst (hunk 1 above)
Comment (profxj) [reply CHANGES.rst:48]: done

File: pypeit/core/wavecal/wvutils.py (hunk 2 above)
Comment (profxj) [reply pypeit/core/wavecal/wvutils.py:11]: I presume you mean `numba` and it appears to be.

File: pypeit/wavecalib.py (hunk 3 above)
Comment (profxj) [reply pypeit/wavecalib.py:122]: pointed to `DataContainer`

File: doc/gemini_gmos.rst (hunk 4 above)
Comment (profxj) [reply doc/gemini_gmos.rst:13]: our sky subtraction is so good, you will likely only increase
the error "doing it right"

**Comment by profxj** [comment]
Unit tests:

================ 253 passed, 377 warnings in 1084.18s (0:18:04) ================

**Comment by profxj** [comment]
(base) profx> ./pypeit_test -t 4 -i gemini_gmos develop
Running tests on the following instruments:
    gemini_gmos

Reducing data from gemini_gmos for the following setups:
    GS_HAM_R400_700
    GS_HAM_R400_860
    GN_HAM_R400_885
    GN_HAM_NS_B600_620

Running tests in 4 parallel processes
 0 passed/ 0 failed/ 0 skipped STARTED gemini_gmos/GS_HAM_R400_860 pypeit (without masters)
 0 passed/ 0 failed/ 0 skipped STARTED gemini_gmos/GN_HAM_R400_885 pypeit (without masters)
 0 passed/ 0 failed/ 0 skipped STARTED gemini_gmos/GS_HAM_R400_700 pypeit (without masters)
 0 passed/ 0 failed/ 0 skipped STARTED gemini_gmos/GN_HAM_NS_B600_620 pypeit (without masters)
 1 passed/ 0 failed/ 0 skipped PASSED  gemini_gmos/GS_HAM_R400_700 pypeit (without masters)
 2 passed/ 0 failed/ 0 skipped PASSED  gemini_gmos/GN_HAM_R400_885 pypeit (without masters)
 3 passed/ 0 failed/ 0 skipped PASSED  gemini_gmos/GS_HAM_R400_860 pypeit (without masters)
 3 passed/ 0 failed/ 0 skipped STARTED gemini_gmos/GS_HAM_R400_860 pypeit standards (without masters)
 4 passed/ 0 failed/ 0 skipped PASSED  gemini_gmos/GS_HAM_R400_860 pypeit standards (without masters)
 4 passed/ 0 failed/ 0 skipped STARTED gemini_gmos/GS_HAM_R400_860 pypeit_sensfunc
 5 passed/ 0 failed/ 0 skipped PASSED  gemini_gmos/GS_HAM_R400_860 pypeit_sensfunc
 5 passed/ 0 failed/ 0 skipped STARTED gemini_gmos/GS_HAM_R400_860 pypeit_flux_setup
 6 passed/ 0 failed/ 0 skipped PASSED  gemini_gmos/GS_HAM_R400_860 pypeit_flux_setup
 6 passed/ 0 failed/ 0 skipped STARTED gemini_gmos/GS_HAM_R400_860 pypeit_flux
 7 passed/ 0 failed/ 0 skipped PASSED  gemini_gmos/GS_HAM_R400_860 pypeit_flux
 7 passed/ 0 failed/ 0 skipped STARTED gemini_gmos/GS_HAM_R400_860 pypeit_coadd_1dspec
 8 passed/ 0 failed/ 0 skipped PASSED  gemini_gmos/GS_HAM_R400_860 pypeit_coadd_1dspec
 9 passed/ 0 failed/ 0 skipped PASSED  gemini_gmos/GN_HAM_NS_B600_620 pypeit (without masters)

--- PYPEIT DEVELOPMENT SUITE PASSED 9/9 TESTS (Masters ignored) ---
Testing Started at 2020-11-13T09:54:28.298387
Testing Completed at 2020-11-13T10:50:01.000779
Total Time: 0:55:32.702392

**Review by jhennawi — APPROVED** [review]

File: doc/gemini_gmos.rst [hunk 5]
```diff
@@ -10,6 +10,14 @@ This file summarizes several instrument specific
...
+===============
+
+For the time being, we have enabled reductions of data
+taken in Nod+Shuffle mode by simply replicating the calibrations.
+That is, we do *not* subtract the nodded images but reduce
+it as if it were a separate slit.
+
```
Comment (deborape) [inline doc/gemini_gmos.rst:20]: I guess this is only for my understanding. In this case can the users subtract the nodded reduced 2D spectra themselves, or will they only be able to coadd the 1D spectra extracted from each frame?

File: pypeit/core/wavecal/spectrographs/templ_gemini_gmos.py [hunk 6]
```diff
@@ -0,0 +1,107 @@
+""" Generate the wavelength templates for Keck/DEIMOS"""
```
Comment (deborape) [inline pypeit/core/wavecal/spectrographs/templ_gemini_gmos.py:1]: "Generate the wavelength templates for Gemini/GMOS"

File: pypeit/core/wavecal/wvutils.py [hunk 7]
```diff
@@ -5,10 +5,13 @@
 .. include:: ../include/links.rst
 
 """
+import os
```
Comment (deborape) [inline pypeit/core/wavecal/wvutils.py:8]: `os` is imported twice.

**Review by deborape — APPROVED** [review]
Looks good to me!

File: doc/gemini_gmos.rst (hunk 5 above)
Comment (profxj) [reply doc/gemini_gmos.rst:20]: The latter, primarily.  Although I guess they'll have the 2D reduced
frames that they could subtract as well.

File: pypeit/core/wavecal/spectrographs/templ_gemini_gmos.py (hunk 6 above)
Comment (profxj) [reply pypeit/core/wavecal/spectrographs/templ_gemini_gmos.py:1]: fixed!

File: pypeit/core/wavecal/wvutils.py (hunk 7 above)
Comment (profxj) [reply pypeit/core/wavecal/wvutils.py:8]: fixed

**Comment by profxj** [comment]
merging

**Pull request merged**